password_enhancements.module

```diff
@@ -0,0 +1,416 @@
+use Drupal\user\UserInterface;
+
+/**
+ * Implements hook_theme().
+ */
+function password_enhancements_theme($existing, $type, $theme, $path): array {
+  return [
+    'password_enhancements_error_messages' => [
+      'variables' => [
+        'error_message' => '',
+        'messages' => [],
+      ],
+    ],
+    'password_enhancements_policy_constraints' => [
+      'variables' => [
+        'constraints' => [],
+        'descriptions' => [],
+        'hidden' => FALSE,
+      ],
+    ],
+    'password_enhancements_logged_out_message' => [
+      'variables' => [
+        'message' => '',
+      ],
+    ],
+  ];
+}
+
+/**
+ * Implements hook_form_FORM_ID_alter().
+ */
+function password_enhancements_form_user_form_alter(&$form, FormStateInterface $form_state, string $form_id): void {
+  password_enhancements_form_user_register_form_alter($form, $form_state, $form_id);
+  $form['password_change_required']['#access'] = Drupal::currentUser()->hasPermission('administer user password settings');
+  array_unshift($form['actions']['submit']['#submit'], 'password_enhancements_form_user_form_alter_submit');
+}
+
+/**
+ * Submit callback for password_enhancements_form_user_form_alter().
+ *
+ * @param array $form
+ *   The form.
+ * @param \Drupal\Core\Form\FormStateInterface $form_state
+ *   The current form state.
+ */
+function password_enhancements_form_user_form_alter_submit(array &$form, FormStateInterface $form_state): void {
+  // Update last updated date for the password.
+  if (!empty($form_state->getValue('pass'))) {
+    $form_state->setValue('password_changed', Drupal::time()->getRequestTime());
+  }
+}
+
+/**
+ * Implements hook_form_FORM_ID_alter().
+ */
+function password_enhancements_form_password_enhancements_change_form_alter(&$form, FormStateInterface $form_state, $form_id) {
+  password_enhancements_form_user_register_form_alter($form, $form_state, $form_id);
+}
+
+/**
+ * Implements hook_form_FORM_ID_alter().
+ */
+function password_enhancements_form_user_register_form_alter(array &$form, FormStateInterface $form_state, string $form_id): void {
+  /** @var \Drupal\Core\Config\Config $config */
+  $config = Drupal::getContainer()->get('config.factory')->get('password_enhancements.settings');
+  if (empty($form['account']['pass']) && !empty($config->get('require_password'))) {
+    $form['account']['pass'] = [
+      '#type' => 'password_confirm',
+      '#size' => 25,
+      '#description' => t('Provide a password for the new account in both fields.'),
+      '#required' => TRUE,
+    ];
+    $form['actions']['submit']['#submit'][] = 'password_enhancements_form_user_register_form_alter_password_submit';
+  }
+  elseif (empty($form['account']['pass'])) {
+    // No password field available, nothing to do here.
+    return;
+  }
+
+  // Add weight for account fields.
+  $weight = 0;
+  foreach ($form['account'] as &$field) {
+    if (is_array($field) && !array_key_exists('#weight', $field)) {
```

Comment:
Did you prove earlier that `isset()` is faster than `array_key_exists()`? :)

Reply:
Yes, but in this case I wanted to check the existence of the array key, because isset() return false if the value is `null`, although I don't know if it would make any sense to set `null` for the `#weight`, so I just wanted to be on the safer side :)